File: 103-binary-tree-zigzag-level-order-traversal/binary-tree-zigzag-level-order-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        if(!root)
            return {};

        vector<vector<int>> res;
        queue<TreeNode*> q;
        q.push(root);
        res.push_back({root->val});
        bool isEven = false;

        while(!q.empty()){
            queue<TreeNode*> q1;
            vector<int> ans;

            while(!q.empty()) {
                TreeNode* node = q.front();
                cout<<node->val<<" ";
                if(node->left) {
                    q1.push(node->left);
                    ans.push_back(node->left->val);
                }
                
                if(node->right){
                    q1.push(node->right);
                    ans.push_back(node->right->val);
                }
                
                q.pop();
            }

            if(!isEven) {
                reverse(ans.begin(), ans.end());
                isEven = true;
            }
            else 
                isEven = false;
            
            if(!q1.empty()) {
                res.push_back(ans);
                q = q1;
            }
        }

        return res;
    }
}; 

File: 103-binary-tree-zigzag-level-order-traversal/binary-tree-zigzag-level-order-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def zigzagLevelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        if root==None:
            return []

        res=[[root.val]]
        q=[root]
        isEven=True

        while len(q):
            que=[]
            ans=[]
            while(len(q)!=0):
                item=q.pop(0)

                if(item.left!=None):
                    que.append(item.left)
                    ans.append(item.left.val)

                if(item.right!=None):
                    que.append(item.right)
                    ans.append(item.right.val)
    
            if isEven:
                ans.reverse()    
                isEven=False
            else:
                isEven=True

            if len(ans):
                res.append(ans)
                q=que  
    
        return res       